# Constants in SciPy
- As SciPy is more focused on scientific implementations, it provides many built-in scientific constants.

- These constants can be helpful when you are working with Data Science.

# Import Library

In [ ]:
import scipy.constants as const
import numpy as np

# Unit

In [ ]:
const.unit('electron volt')

'J'

# Value

In [ ]:
const.value('electron volt')

1.602176634e-19

# Precision

In [ ]:
const.precision('alpha particle mass')

3.1604338507842195e-10

# Find

In [ ]:
const.find('electron volt')

['atomic mass unit-electron volt relationship',
 'electron volt',
 'electron volt-atomic mass unit relationship',
 'electron volt-hartree relationship',
 'electron volt-hertz relationship',
 'electron volt-inverse meter relationship',
 'electron volt-joule relationship',
 'electron volt-kelvin relationship',
 'electron volt-kilogram relationship',
 'hartree-electron volt relationship',
 'hertz-electron volt relationship',
 'inverse meter-electron volt relationship',
 'joule-electron volt relationship',
 'kelvin-electron volt relationship',
 'kilogram-electron volt relationship']

# Mathematical constants
Physical and mathematical constants and units.

In [ ]:
print(const.pi)
print(const.golden)
print(const.golden_ratio)

3.141592653589793
1.618033988749895
1.618033988749895


# Physical Constants
The following physical constants are available as attributes of [scipy.constants](https://docs.scipy.org/doc/scipy/reference/constants.html#module-scipy.constants). All units are `SI`.

---

## ⚛️ Fundamental Physical Constants

| **Attribute**            | **Quantity / Description**                | **Units**   |
| ------------------------ | ----------------------------------------- | ----------- |
| `c`                      | speed of light in vacuum                  | m s⁻¹       |
| `speed_of_light`         | speed of light in vacuum                  | m s⁻¹       |
| `mu_0`                   | magnetic constant                         | N A⁻²       |
| `epsilon_0`              | electric constant (vacuum permittivity)   | F m⁻¹       |
| `h`                      | Planck constant                           | J Hz⁻¹      |
| `Planck`                 | Planck constant                           | J Hz⁻¹      |
| `hbar`                   | reduced Planck constant                   | J s         |
| `G`                      | Newtonian constant of gravitation         | m³ kg⁻¹ s⁻² |
| `gravitational_constant` | Newtonian constant of gravitation         | m³ kg⁻¹ s⁻² |
| `g`                      | standard acceleration of gravity          | m s⁻²       |
| `e`                      | elementary charge                         | C           |
| `elementary_charge`      | elementary charge                         | C           |
| `R`                      | molar gas constant                        | J mol⁻¹ K⁻¹ |
| `gas_constant`           | molar gas constant                        | J mol⁻¹ K⁻¹ |
| `alpha`                  | fine-structure constant                   | (unitless)  |
| `fine_structure`         | fine-structure constant                   | (unitless)  |
| `N_A`                    | Avogadro constant                         | mol⁻¹       |
| `Avogadro`               | Avogadro constant                         | mol⁻¹       |
| `k`                      | Boltzmann constant                        | J K⁻¹       |
| `Boltzmann`              | Boltzmann constant                        | J K⁻¹       |
| `sigma`                  | Stefan–Boltzmann constant                 | W m⁻² K⁻⁴   |
| `Stefan_Boltzmann`       | Stefan–Boltzmann constant                 | W m⁻² K⁻⁴   |
| `Wien`                   | Wien wavelength displacement law constant | m K         |
| `Rydberg`                | Rydberg constant                          | m⁻¹         |
| `m_e`                    | electron mass                             | kg          |
| `electron_mass`          | electron mass                             | kg          |
| `m_p`                    | proton mass                               | kg          |
| `proton_mass`            | proton mass                               | kg          |
| `m_n`                    | neutron mass                              | kg          |
| `neutron_mass`           | neutron mass                              | kg          |

---


In [ ]:
print(const.c)
print(const.speed_of_light)
print(const.mu_0)
print(const.epsilon_0)
print(const.Planck)
print(const.hbar)
print(const.G)
print(const.g)
print(const.sigma)

299792458.0
299792458.0
1.25663706127e-06
8.8541878188e-12
6.62607015e-34
1.0545718176461565e-34
6.6743e-11
9.80665
5.6703744191844314e-08


# Constant Units
A list of all units under the constants module can be seen using the `dir()` function

# Unit Categories
The units are placed under these categories:

- Metric
- Binary
- Mass
- Angle
- Time
- Length
- Pressure
- Volume
- Speed
- Temperature
- Energy
- Power
- Force

In [ ]:
dir(const)

['Avogadro',
 'Boltzmann',
 'Btu',
 'Btu_IT',
 'Btu_th',
 'ConstantWarning',
 'G',
 'Julian_year',
 'N_A',
 'Planck',
 'R',
 'Rydberg',
 'Stefan_Boltzmann',
 'Wien',
 '__all__',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '_codata',
 '_constants',
 '_obsolete_constants',
 'acre',
 'alpha',
 'angstrom',
 'arcmin',
 'arcminute',
 'arcsec',
 'arcsecond',
 'astronomical_unit',
 'atm',
 'atmosphere',
 'atomic_mass',
 'atto',
 'au',
 'bar',
 'barrel',
 'bbl',
 'blob',
 'c',
 'calorie',
 'calorie_IT',
 'calorie_th',
 'carat',
 'centi',
 'codata',
 'constants',
 'convert_temperature',
 'day',
 'deci',
 'degree',
 'degree_Fahrenheit',
 'deka',
 'dyn',
 'dyne',
 'e',
 'eV',
 'electron_mass',
 'electron_volt',
 'elementary_charge',
 'epsilon_0',
 'erg',
 'exa',
 'exbi',
 'femto',
 'fermi',
 'find',
 'fine_structure',
 'fluid_ounce',
 'fluid_ounce_US',
 'fluid_ounce_imp',
 'foot',
 'g',
 'gallon',
 'gallon_US',
 'g

# Metrics

Return the specified unit in `meter` (e.g. centi returns 0.01)

In [ ]:
# print(dir(const.zetta))
print(const.quetta)
print(const.ronna)
print(const.yotta)
print(const.zetta)
print(const.peta)
print(const.tera)
print(const.giga)
print(const.mega)
print(const.kilo)
print(const.hecto)
print(const.deka)
print(const.deci)
print(const.centi)
print(const.milli)
print(const.micro)
print(const.nano)
print(const.pico)
print(const.femto)
print(const.atto)
print(const.zepto)
print(const.yocto)
print(const.ronto)
print(const.quecto)

1e+30
1e+27
1e+24
1e+21
1000000000000000.0
1000000000000.0
1000000000.0
1000000.0
1000.0
100.0
10.0
0.1
0.01
0.001
1e-06
1e-09
1e-12
1e-15
1e-18
1e-21
1e-24
1e-27
1e-30


# Binary
Return the specified unit in `bytes` (e.g. kibi returns 1024)

In [ ]:
print(const.kibi)
print(const.mebi)
print(const.gibi)
print(const.tebi)
print(const.pebi)
print(const.exbi)
print(const.zebi)
print(const.yobi)

1024
1048576
1073741824
1099511627776
1125899906842624
1152921504606846976
1180591620717411303424
1208925819614629174706176


# Mas

Return the specified unit in `kg` (e.g. gram returns 0.001)

In [ ]:
print(const.gram)
print(const.metric_ton)
print(const.grain)
print(const.lb)
print(const.oz)
print(const.ounce)
print(const.stone)
print(const.long_ton)
print(const.short_ton)
print(const.troy_ounce)
print(const.troy_pound)
print(const.carat)
print(const.atomic_mass)
print(const.m_u)
print(const.u)

0.001
1000.0
6.479891e-05
0.45359236999999997
0.028349523124999998
0.028349523124999998
6.3502931799999995
1016.0469088
907.1847399999999
0.031103476799999998
0.37324172159999996
0.0002
1.66053906892e-27
1.66053906892e-27
1.66053906892e-27


# Angle

Return the specified unit in `radians` (e.g. degree returns 0.017453292519943295)

In [ ]:
print(const.degree)
print(const.arcmin)
print(const.arcminute)
print(const.arcsec)
print(const.arcsecond)

0.017453292519943295
0.0002908882086657216
0.0002908882086657216
4.84813681109536e-06
4.84813681109536e-06


# Pressure:
Return the specified unit in `pascals` (e.g. psi returns 6894.757293168361)

In [ ]:
print(const.atm)
print(const.atmosphere)
print(const.bar)
print(const.torr)
print(const.mmHg)
print(const.psi)

101325.0
101325.0
100000.0
133.32236842105263
133.32236842105263
6894.757293168361


# Area:
Return the specified unit in `square meters`(e.g. hectare returns 10000.0)

In [ ]:
print(const.hectare)
print(const.acre)

10000.0
4046.8564223999992


# Volume:
Return the specified unit in `cubic meters` (e.g. liter returns 0.001)

In [ ]:
print(const.liter)
print(const.litre)
print(const.gallon)
print(const.gallon_US)
print(const.gallon_imp)
print(const.fluid_ounce)
print(const.fluid_ounce_US)
print(const.fluid_ounce_imp)
print(const.barrel)
print(const.bbl)

0.001
0.001
0.0037854117839999997
0.0037854117839999997
0.00454609
2.9573529562499998e-05
2.9573529562499998e-05
2.84130625e-05
0.15898729492799998
0.15898729492799998


# Speed:
Return the specified unit in `meters per second` (e.g. speed_of_sound returns 340.5)

In [ ]:
print(const.kmh)
print(const.mph)
print(const.mach)
print(const.speed_of_sound)
print(const.knot)

0.2777777777777778
0.44703999999999994
340.5
340.5
0.5144444444444445


# Temperature:
Return the specified unit in `Kelvin` (e.g. zero_Celsius returns 273.15)

In [ ]:
print(const.zero_Celsius)
print(const.degree_Fahrenheit)

273.15
0.5555555555555556


# Energy:
Return the specified unit in `joules` (e.g. calorie returns 4.184)

In [ ]:
print(const.calorie)
print(const.calorie_th)
print(const.calorie_IT)
print(const.erg)
print(const.Btu)
print(const.Btu_IT)
print(const.Btu_th)
print(const.ton_TNT)
print(const.eV)

4.184
4.184
4.1868
1e-07
1055.05585262
1055.05585262
1054.3502644888888
4184000000.0
1.602176634e-19


# Power:
Return the specified unit in `watts` (e.g. horsepower returns 745.6998715822701)

In [ ]:
print(const.hp)
print(const.horsepower)

745.6998715822701
745.6998715822701


# Force:
Return the specified unit in `newton` (e.g. kilogram_force returns 9.80665)

In [ ]:
print(const.dyn)
print(const.dyne)
print(const.lbf)
print(const.pound_force)
print(const.kgf)
print(const.kilogram_force)

1e-05
1e-05
4.4482216152605
4.4482216152605
9.80665
9.80665


# Time

In [ ]:
print(const.minute)
print(const.hour)
print(const.day)
print(const.week)
print(const.year)
print(const.Julian_year)

60.0
3600.0
86400.0
604800.0
31536000.0
31557600.0


# Length

In [ ]:
print(const.inch)
print(const.foot)
print(const.yard)
print(const.mile)
print(const.mil)
print(const.pt)
print(const.point)
print(const.survey_foot)
print(const.survey_mile)
print(const.fermi)
print(const.angstrom)
print(const.micron)
print(const.au)
print(const.astronomical_unit)
print(const.light_year)
print(const.parsec)

0.0254
0.30479999999999996
0.9143999999999999
1609.3439999999998
2.5399999999999997e-05
0.00035277777777777776
0.00035277777777777776
0.3048006096012192
1609.3472186944373
1e-15
1e-10
1e-06
149597870700.0
149597870700.0
9460730472580800.0
3.085677581491367e+16


# convert_temperature

In [ ]:
const.convert_temperature(35.8, 'c', 'F')

np.float64(96.44000000000003)

In [ ]:
const.convert_temperature(96.44, 'fahrenheit', 'Celsius')

np.float64(35.80000000000001)

In [ ]:
const.convert_temperature(np.array([35.8, 99, 180]), 'F', 'K')

array([275.26111111, 310.37222222, 355.37222222])

# lambda2nu
```python
def lambda2nu(lambda_):
   # c is speed of light(m/s)
   c = 3e8  
   # Convert wavelength (in meters) to frequency (in Hz)
   nu = c / lambda_
   return nu

# w is wavelength(m)
w = 500e-9
f = lambda2nu(w)
print(f"Frequency: {f} Hz")
```

In [ ]:
const.lambda2nu(500e-9)

np.float64(599584916000000.0)